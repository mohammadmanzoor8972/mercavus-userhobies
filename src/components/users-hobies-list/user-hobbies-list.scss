@import '../../styles/main';

.container {
  font-size: $default-font-size;

  &-empty {
    padding: 1rem;
    justify-content: center;
    align-items: center;
    display: flex;
  }
}

.item {
  display: grid;
  grid-template-columns: repeat(3, 1fr) 50px;
  font-size: $default-font-size;
  grid-column-gap: 5rem;
  padding: 1rem;
  
  border: 1px solid $black;
  border-bottom: 0px;

  &:last-child {
    border-top: 1px solid $black;
    border-bottom: 1px solid $black;
    border: 1px solid $black;
  }

  &:first-child {
    border-top: 0px;
  }

  &:hover {
    background: $grey;
  }
}

.delete-btn {
  text-align: center;
  cursor: pointer;
  &:hover {
    color: $red;
  }
}
