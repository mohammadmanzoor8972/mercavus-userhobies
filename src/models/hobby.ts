import { PassionLevel } from "./passion";

  export interface Hobby {
    name: string
    level: PassionLevel | string
    date: string
  }
  
  