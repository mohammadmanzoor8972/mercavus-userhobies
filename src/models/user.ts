import { Hobby } from './hobby'

export interface User {
  id: string
  name: string
  hobbies: Hobby[]
}
