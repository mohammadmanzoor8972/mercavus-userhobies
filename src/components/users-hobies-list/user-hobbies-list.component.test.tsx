import * as React from 'react'
import { mount } from 'enzyme'
import { UserHobbiesList } from './user-hobbies-list.component'
import { User } from '../../models/user'
import { PassionLevel } from '../../models/passion'

describe('UserHobbiesList component', () => {
  const mockUser: User = {
    id: '123',
    name: 'mock',
    hobbies: [],
  }

  it('should render no hobbies placeholder when no hobbies defined', () => {
    expect(
      mount(<UserHobbiesList user={mockUser} hobbies={mockUser.hobbies} removeHobby={jest.fn()} />)
    ).toMatchSnapshot()
  })

  it('should render hobbies', () => {
    expect(
      mount(<UserHobbiesList user={mockUser} hobbies={[{
        name: 'hobby 1',
        date: 'since now',
        level: PassionLevel.High,
      }, {
        name: 'hobby 2',
        date: 'since big bang',
        level: PassionLevel.VeryHigh,
      }]} removeHobby={jest.fn()} />)
    ).toMatchSnapshot()
  })

  it('should show remove confirmation message', () => {
    window.confirm = jest.fn()
    const removeMock = jest.fn()
    const sut = mount(<UserHobbiesList user={mockUser} hobbies={[{
      name: 'hobby 1',
      date: 'since now',
      level: PassionLevel.High,
    }]} removeHobby={removeMock} />)

    sut.findWhere(x => x.text() === 'x').at(0).simulate('click')
    expect(removeMock).not.toHaveBeenCalled()
  })

  it('should remove hobby on remove button click if user confirms the action ', () => {
    window.confirm = jest.fn(() => true)
    const removeMock = jest.fn()
    const mockHobby = {
      name: 'hobby 1',
      date: 'since now',
      level: PassionLevel.High,
    }
    const sut = mount(<UserHobbiesList user={mockUser} hobbies={[mockHobby]} removeHobby={removeMock} />)

    sut.findWhere(x => x.text() === 'x').at(0).simulate('click')
    expect(removeMock).toHaveBeenCalledWith({user: mockUser, hobby: mockHobby})
  })
})