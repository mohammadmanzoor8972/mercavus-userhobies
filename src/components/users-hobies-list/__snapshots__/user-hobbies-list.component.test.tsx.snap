// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserHobbiesList component should render hobbies 1`] = `
<Component
  hobbies={
    Array [
      Object {
        "date": "since now",
        "level": "High",
        "name": "hobby 1",
      },
      Object {
        "date": "since big bang",
        "level": "Very-High",
        "name": "hobby 2",
      },
    ]
  }
  removeHobby={[MockFunction]}
  user={
    Object {
      "hobbies": Array [],
      "id": "123",
      "name": "mock",
    }
  }
>
  <div
    className="container"
  >
    <div
      className="item"
      key="hobby 1-High"
    >
      <span
        className="level"
      >
        High
      </span>
      <span
        className="name"
      >
        hobby 1
      </span>
      <span
        className="date"
      >
        since now
      </span>
      <button
        className="deleteBtn"
        onClick={[Function]}
      >
        x
      </button>
    </div>
    <div
      className="item"
      key="hobby 2-Very-High"
    >
      <span
        className="level"
      >
        Very-High
      </span>
      <span
        className="name"
      >
        hobby 2
      </span>
      <span
        className="date"
      >
        since big bang
      </span>
      <button
        className="deleteBtn"
        onClick={[Function]}
      >
        x
      </button>
    </div>
  </div>
</Component>
`;

exports[`UserHobbiesList component should render no hobbies placeholder when no hobbies defined 1`] = `
<Component
  hobbies={Array []}
  removeHobby={[MockFunction]}
  user={
    Object {
      "hobbies": Array [],
      "id": "123",
      "name": "mock",
    }
  }
>
  <div
    className="container containerEmpty"
  >
    <NoHobbiesPlaceholder
      user={
        Object {
          "hobbies": Array [],
          "id": "123",
          "name": "mock",
        }
      }
    >
      <span>
        mock has no specified his hobbies yet. Press add button above to add first hobby!
      </span>
    </NoHobbiesPlaceholder>
  </div>
</Component>
`;
